module/applications/activity/activity-usage-dialog.mjs

```diff
@@ -1,11 +1,411 @@
+        selected: (isArray && this.config.consume.resources.includes(value))
+          || (!isArray && (this.config.consume?.resources !== false) && (this.config.consume !== false))
+      };
+    });
+    if ( resources.length ) context.resources = {
+      field: new SetField(new NumberField()),
+      name: "consume.resources",
+      options: resources,
+      dataset: { dtype: "Number" }
+    };
+
+    context.hasConsumption = context.spellSlot || context.resources;
+
+    // TODO: Add improved consumption hints once consumption is implemented
+
+    return context;
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Prepare rendering context for the creation section.
+   * @param {ApplicationRenderContext} context  Context being prepared.
+   * @param {HandlebarsRenderOptions} options   Options which configure application rendering behavior.
+   * @returns {Promise<ApplicationRenderContext>}
+   * @protected
+   */
+  async _prepareCreationContext(context, options) {
+    context.hasCreation = false;
+    if ( this.activity.target.template.type ) {
+      context.hasCreation = true;
+      context.template = {
+        field: new BooleanField({ label: game.i18n.localize("DND5E.PlaceTemplate") }),
+        name: "create.measuredTemplate",
+        value: this.config.create?.measuredTemplate
+      };
+    }
+    return context;
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Prepare rendering context for the footer.
+   * @param {ApplicationRenderContext} context  Context being prepared.
+   * @param {HandlebarsRenderOptions} options   Options which configure application rendering behavior.
+   * @returns {Promise<ApplicationRenderContext>}
+   * @protected
+   */
+  async _prepareFooterContext(context, options) {
+    context.buttons = [{
+      action: "use",
+      icon: `fa-solid fa-${this.activity.isSpell ? "magic" : "fist-raised"}`,
+      label: `DND5E.AbilityUse${this.activity.isSpell ? "Cast" : "Use"}`,
+      type: "button"
+    }];
+    return context;
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Prepare rendering context for the scaling section.
+   * @param {ApplicationRenderContext} context  Context being prepared.
+   * @param {HandlebarsRenderOptions} options   Options which configure application rendering behavior.
+   * @returns {Promise<ApplicationRenderContext>}
+   * @protected
+   */
+  async _prepareScalingContext(context, options) {
+    context.hasScaling = true;
+    context.notes = [];
+
+    if ( this.actor.system.spells && this.activity.isSpell && (this.item.system.level > 0) ) {
+      const minimumLevel = this.item.system.level ?? 1;
+      const maximumLevel = Object.values(this.actor.system.spells)
+        .reduce((max, d) => d.max ? Math.max(max, d.level) : max, 0);
+
+      const spellSlotOptions = Object.entries(this.actor.system.spells).map(([value, slot]) => {
```

Comment:
We should probably use the same logic as `AbilityUseDialog._createSpellSlotOptions` (or even call into that method since it's static) in order to make sure we pick up any alternative casting options.

Reply:
Maybe I'll have @krbz999 check this, but I believe this should continue working with alt types because all alt types should exist in `actor.system.spells`. The old code looped over spell levels and alt spellcasting types and then fetched the data from `system.spells`, while this loops over `system.spells` and then uses the data there directly.

I did make a change to `prepareAltSlots` so it places the `type` property into each slot so the system can track it back to a specific spellcasting type.

I just did a tweak to hopefully keep the label more in line with the old implementation.

Reply:
Sounds good to me then. Let's try it out.

Reply:
> Maybe I'll have @krbz999 check this, but I believe this should continue working with alt types because all alt types should exist in `actor.system.spells`. The old code looped over spell levels and alt spellcasting types and then fetched the data from `system.spells`, while this loops over `system.spells` and then uses the data there directly.

Yeah that sounds about right. Nothing is ever deleted from the `spells` object, so I believe we checked for the existence of a `.levels` property.